File: src/maven.js
const path = require('path')
const exec = require('./exec')
const { MAVEN_LIST, MAVEN_PRE_RUN } = require('./paths')
const mappers = require('../bin/maven/maven-mappers.json')

module.exports.list = async (projectRootPath) => {
  const pomPath = path.join(projectRootPath, 'pom.xml')
  const stdOut = await exec(MAVEN_LIST, pomPath)
  const dependencies = extractDependencies(stdOut)
  return useMapper(dependencies)
}

module.exports.preRun = async (projectRootPath) => {
  const pomPath = path.join(projectRootPath, 'pom.xml')
  const stdOut = await exec(MAVEN_PRE_RUN, pomPath)
  const dependencies = extractDependencies(stdOut)
  return useMapper(dependencies)
}

const extractDependencies = (stdOut) => stdOut.split(/\r\n|\n/)
  .map(line => line.split('--')[0].trim())
  .filter(dependency => !dependency.endsWith('test'))
  .map(dependency => {
    const [groupId, artifactId, pkg, version, scope] = dependency.split(':')
    return {
      artifactId,
      groupId,
      version
    }
  })

const useMapper = (dependencies) => {
  return dependencies.map(mapDependency)
}

const mapDependency = (dependency) => {

  for (const { groupId, artifactId, cpe } of mappers) {
    if (groupId === dependency.groupId && (artifactId === '*' || artifactId === dependency.artifactId)) {
      // console.debug(`Override ${dependency.artifactId} with ${cpe}.`)
      return {
        product: cpe,
        version: dependency.version
      }
    }
  }
  return {
    product: dependency.artifactId,
    version: dependency.version
  }
}

File: vscnc/maven.py
from vscnc.runner import run


def pre_run(project_root_dir):
    pom_path = f'{project_root_dir}/pom.xml'
    str(run(['bin/maven/preRun.sh', pom_path]))


def list_dependencies(project_root_dir) -> list:
    pom_path = f'{project_root_dir}/pom.xml'
    std_out = run(['bin/maven/list.sh', pom_path]).decode('ascii')
    return _extract_dependencies(std_out)


def _extract_dependencies(std_out: str) -> list:
    lines = std_out.splitlines()
    raw_dependencies = map(lambda line: line.split('--')[0], lines)
    filtered_dependencies = filter(lambda line: not line.endswith('test'), raw_dependencies)

    return list(map(_map_dependency, filtered_dependencies))


def _map_dependency(dependency: str):
    artifactId = None
    dependency_parts = dependency.split(':')

    if len(dependency_parts) in [4, 5]:
        artifactId = dependency_parts[1]
        version = dependency_parts[3]

    if len(dependency_parts) == 6:
        artifactId = dependency_parts[1]
        version = dependency_parts[4]

    return {
        'product': artifactId.strip(),
        'version': version.strip()
    }
